### PR TITLE
Add step by step introduction/tutorial to question pages 

### LINKED ISSUE #51: Add tutorial for test case based programming
It appears users are struggling with understanding the test case nature of codeWOF questions. A tutorial should help ease them into the system, possibly interactive for a question.

See https://introjs.com/

### PR DESCRIPTION
**Pull request by courtneycb**
**Changes**

- Add button to the top right of all question pages that triggers the intro/tutorial when clicked.
- Add step by step explanations for each main component of the question page. These are different for each question type (program, function, debugging and parsons).
- Hides the tutorial button on devices that are of medium size and smaller. I decided to do this because introjs doesn't automatically scroll to show elements if they are hidden inside a scrollable div. For example if the test status was cut off on a mobile device, when the tutorial highlights this section it will be out of the viewport.
- Add licence files for introjs. It appears we have no other third party licences currently, these should be added in a separate PR.
- Uses the 'modern' introjs theme for tooltips.

Resolves #51

### CODE REVIEW

**Comment by eAlasdair** [comment]
> It appears we have no other third party licences currently, these should be added in a separate PR.

The hek happened to #271 ? :confounded:

**Comment by eAlasdair** [comment]
> The hek happened to #271 ? confounded

Okay I don't know how (I 100% checked this before making the PR for 271) , but all the licences are in `codewof/codewof/` rather than `codewof/`. Could you please move them over as part of this?

File: codewof/static/css/third-party/introjs.css [hunk 1]
```diff
@@ -0,0 +1,506 @@
+.introjs-overlay {
```
Comment (eAlasdair) [inline codewof/static/css/third-party/introjs.css:1]: Does this sentence in our README apply regarding this file (and/or the other one)?

> If a source file of a third-party library or system is stored within this repository, then a copyright notice should be present at the top of the file.

**Comment by courtneycb** [comment]
Review changes made. 

There are cases where a scrollable div will show even on large screens and introjs will highlight sections that are out of the viewport. Unsure what should be done about this.
image
image

**Comment by eAlasdair** [comment]
> There are cases where a scrollable div will show even on large screens and introjs will highlight sections that are out of the viewport. Unsure what should be done about this.

Can we set the position of the scrollbar (if it exists) at the time of the highlight?

Else we might need to use screenshots or something

**Comment by courtneycb** [comment]
> > There are cases where a scrollable div will show even on large screens and introjs will highlight sections that are out of the viewport. Unsure what should be done about this.
> 
> Can we set the position of the scrollbar (if it exists) at the time of the highlight?
> 
> Else we might need to use screenshots or something

I can, but introjs grabs the elements position before it scrolls, so it highlights where it used to be. I've tried a lot of things but keep running into very finicky situations like this.

image

File: codewof/static/js/question_types/debugging.js [hunk 2]
```diff
@@ -152,3 +158,56 @@ function run_python_code(user_code, test_case) {
...
+
+
+function setTutorialAttributes() {
+    $(".question-text").attr(
+        'data-intro',
+        'This is a description of what the code should do. In debugging questions there will be one or more bugs in the given code that you have to find and correct in order for all the tests to pass.'
+    );
+    $("#python-editor").attr(
+        'data-intro',
+        "This is the code that contains the bug(s). Find the bug(s) and correct them. The greyed out lines cannot be edited."
```
Comment (eAlasdair) [inline codewof/static/js/question_types/debugging.js:170]:
```suggestion
        "This is the code that contains the bug(s). Find the bug(s) and correct them. Greyed out lines cannot be edited."
```

File: codewof/static/js/question_types/debugging.js [hunk 3]
```diff
@@ -152,3 +158,56 @@ function run_python_code(user_code, test_case) {
...
+    );
+    $("#run_code").attr(
+        'data-intro',
+        "Click 'Run code' to run your code against the test cases and see if you fixed the bug!"
```
Comment (eAlasdair) [inline codewof/static/js/question_types/debugging.js:174]: Maybe reword this so it isn't an instruction - people might actually click the button and get confused when nothing happens

File: codewof/static/js/question_types/debugging.js [hunk 4]
```diff
@@ -152,3 +158,56 @@ function run_python_code(user_code, test_case) {
...
+    );
+    // the input for the first test case
+    $('#test-case-table tbody tr:nth-child(1) td:eq(0)').attr(
+        'data-intro',
+        'This is the test code that has been run for this particular test.'
+    );
+    // the expected output for the first test case
+    $('#test-case-table tbody tr:nth-child(1) td:eq(1)').attr(
+        'data-intro',
+        'This is the output that the test case expects to be printed for the given input.'
```
Comment (eAlasdair) [inline codewof/static/js/question_types/debugging.js:197]:
```suggestion
        'This is the output that the test code is expected to print.'
```

File: codewof/static/js/question_types/debugging.js [hunk 5]
```diff
@@ -152,3 +158,56 @@ function run_python_code(user_code, test_case) {
...
+    );
+    // the received output for the first test case
+    $('#test-case-table tbody tr:nth-child(1) td:eq(2)').attr(
+        'data-intro',
+        'This is the output that has been printed by the test case.\
```
Comment (eAlasdair) [inline codewof/static/js/question_types/debugging.js:202]:
```suggestion
        'This is the output that has been printed by the test code.\
```

File: codewof/static/js/question_types/debugging.js [hunk 6]
```diff
@@ -152,3 +158,56 @@ function run_python_code(user_code, test_case) {
...
+        The test case either prints the return value of the code, or the code executed by the test prints it itself.'
```
Comment (eAlasdair) [inline codewof/static/js/question_types/debugging.js:203]: I don't think we need this (it's also not grammatically correct, I just wrote it to explain my reasoning)

File: codewof/static/js/question_types/debugging.js [hunk 7]
```diff
@@ -152,3 +158,56 @@ function run_python_code(user_code, test_case) {
...
+    );
+    // the status of the first test case
+    $('#test-case-table tbody tr:nth-child(1) td:eq(3)').attr(
+        'data-intro',
+        "This is the status of the test. It will say one of three things: 'Not yet run', 'Passed' or 'Failed'.\
+        'Not yet run' means your code has not yet been run against the test case.\
+        'Passed' means the received output matched the expected output.\
+        'Failed' means the received output did not match the expected output."
```
Comment (eAlasdair) [inline codewof/static/js/question_types/debugging.js:211]: I still think it could be as simple as "A test case will pass if the received output matches the expected output. If all test cases pass the question has been solved." or something similar. Those three terms are self-explanatory in this context

File: codewof/static/js/question_types/function.js [hunk 8]
```diff
@@ -108,3 +114,53 @@ function run_python_code(user_code, test_case) {
...
+    );
+    // the first row in the test case table
+    $('#test-case-table tbody tr:nth-child(1)').attr(
+        'data-intro',
+        'Here is the first test case.'
+    );
+    // the input for the first test case
+    $('#test-case-table tbody tr:nth-child(1) td:eq(0)').attr(
+        'data-intro',
+        'This is how your function will be called. Pay close attention to the input that is passed to the function.'
```
Comment (eAlasdair) [inline codewof/static/js/question_types/function.js:145]: It's given the term 'test code' in the debugging helper so maybe be consistent here

"This is the test code that is run for this particular test. Pay close ..."

File: codewof/static/js/question_types/function.js [hunk 9]
```diff
@@ -108,3 +114,53 @@ function run_python_code(user_code, test_case) {
...
+    );
+    // the expected output for the first test case
+    $('#test-case-table tbody tr:nth-child(1) td:eq(1)').attr(
+        'data-intro',
+        'This is the output that the test case expects to be printed for the given input.'
```
Comment (eAlasdair) [inline codewof/static/js/question_types/function.js:150]:
```suggestion
        'This is the output that the test code is expected to print.'
```

File: codewof/static/js/question_types/parsons.js [hunk 10]
```diff
@@ -122,3 +128,57 @@ function run_python_code(user_code, test_case) {
...
+    );
+    // the first row in the test case table
+    $('#test-case-table tbody tr:nth-child(1)').attr(
+        'data-intro',
+        'Here is the first test case.'
+    );
+    // the input for the first test case
+    $('#test-case-table tbody tr:nth-child(1) td:eq(0)').attr(
+        'data-intro',
+        'This is the test code that has been run for this particular test.'
```
Comment (eAlasdair) [inline codewof/static/js/question_types/parsons.js:163]:
```suggestion
        'This is the test code that is run for this particular test.'
```

File: codewof/static/js/question_types/parsons.js [hunk 11]
```diff
@@ -122,3 +128,57 @@ function run_python_code(user_code, test_case) {
...
+    );
+    // the expected output for the first test case
+    $('#test-case-table tbody tr:nth-child(1) td:eq(1)').attr(
+        'data-intro',
+        'This is the output that the test case expects to be printed for the given input.'
```
Comment (eAlasdair) [inline codewof/static/js/question_types/parsons.js:168]:
```suggestion
        'This is the output that the test code is expected to print.'
```

File: codewof/static/js/question_types/function.js [hunk 12]
```diff
@@ -108,3 +114,53 @@ function run_python_code(user_code, test_case) {
...
+    );
+    // the received output for the first test case
+    $('#test-case-table tbody tr:nth-child(1) td:eq(2)').attr(
+        'data-intro',
+        'This is the output that has been printed by the test case.\
```
Comment (eAlasdair) [inline codewof/static/js/question_types/function.js:155]:
```suggestion
        'This is the output that has been printed by the test code.\
```

File: codewof/static/js/question_types/parsons.js [hunk 13]
```diff
@@ -122,3 +128,57 @@ function run_python_code(user_code, test_case) {
...
+    );
+    // the received output for the first test case
+    $('#test-case-table tbody tr:nth-child(1) td:eq(2)').attr(
+        'data-intro',
+        'This is the output that has been printed by the test case.\
```
Comment (eAlasdair) [inline codewof/static/js/question_types/parsons.js:173]:
```suggestion
        'This is the output that has been printed by the test code.\
```

File: codewof/static/js/question_types/program.js [hunk 14]
```diff
@@ -112,3 +118,51 @@ function run_python_code(user_code, test_case) {
...
+    );
+    // the first row in the test case table
+    $('#test-case-table tbody tr:nth-child(1)').attr(
+        'data-intro',
+        'Here is the first test case.'
+    );
+    // the input for the first test case
+    $('#test-case-table tbody tr:nth-child(1) td:eq(0)').attr(
+        'data-intro',
+        'Each line will be passed to your program as input for this particular test.'
```
Comment (eAlasdair) [inline codewof/static/js/question_types/program.js:148]: Maybe "If your program asks a user for input, each of these lines will be given in place of a real user." Not confident about the exact wording but that sort of style

File: codewof/static/js/question_types/function.js [hunk 15]
```diff
@@ -108,3 +114,53 @@ function run_python_code(user_code, test_case) {
         test_case.runtime_error = true;
     }
 }
+
+
+function setTutorialAttributes() {
+    $(".question-text").attr(
+        'data-intro',
+        'This is a description of what the function should do. It will tell you what your function should return (or print) in order to pass the tests.\
+        Make sure to pay close attention to whether your function should return or print a value!'
```
Comment (eAlasdair) [inline codewof/static/js/question_types/function.js:123]:
```suggestion
        Pay close attention to whether your function should return or print a value!'
```

File: codewof/static/js/question_types/parsons.js [hunk 16]
```diff
@@ -122,3 +128,57 @@ function run_python_code(user_code, test_case) {
         test_case.runtime_error = true;
     }
 }
+
+
+function setTutorialAttributes() {
+    $(".question-text").attr(
+        'data-intro',
+        'This is a description of what the code should do. It will tell you what your code should return (or print) for certain inputs.\
+        Make sure you pay close attention to whether your code should return or print a value!'
```
Comment (eAlasdair) [inline codewof/static/js/question_types/parsons.js:137]:
```suggestion
        Pay close attention to whether your code should return or print a value!'
```

**Review by eAlasdair — COMMENTED** [review]
I think the wording is still a bit iffy, particularly around test cases

**Review by eAlasdair — APPROVED** [review]
I've been told my changes are good.

Nice work!
